android/app/src/main/java/org/jitsi/meet/AudioModeModule.java

```diff
@@ -0,0 +1,440 @@
+            setupAudioRouteChangeDetectionOld();
+        }
+    }
+
+    /**
+     * Audio route change detection mechanism for Android API >= 23.
+     */
+    @TargetApi(Build.VERSION_CODES.M)
+    private void setupAudioRouteChangeDetectionM() {
+        android.media.AudioDeviceCallback audioDeviceCallback =
+                new android.media.AudioDeviceCallback() {
+                    public void onAudioDevicesAdded(AudioDeviceInfo[] addedDevices) {
+                        Log.d(TAG, "Audio devices added");
+                        onAudioDeviceChange();
+                    }
+
+                    public void onAudioDevicesRemoved(AudioDeviceInfo[] removedDevices) {
+                        Log.d(TAG, "Audio devices removed");
+                        onAudioDeviceChange();
+                    }
+                };
+
+        audioManager.registerAudioDeviceCallback(audioDeviceCallback, null);
+    }
+
+    /**
+     * Audio route change detection mechanism for Android API < 23.
+     */
+    private void setupAudioRouteChangeDetectionOld() {
+        ReactContext reactContext = getReactApplicationContext();
+        if (reactContext == null) {
+            return;
+        }
+
+        // Detect changes in wired headset connections
+        final String ACTION_HEADSET_PLUG = (android.os.Build.VERSION.SDK_INT >= 21) ?
+                AudioManager.ACTION_HEADSET_PLUG : Intent.ACTION_HEADSET_PLUG;
+        IntentFilter wiredHeadSetFilter = new IntentFilter(ACTION_HEADSET_PLUG);
+        BroadcastReceiver wiredHeadsetReceiver = new BroadcastReceiver() {
+            @Override
+            public void onReceive(Context context, Intent intent) {
+                Log.d(TAG, "Wired headset added / removed");
+                onAudioDeviceChange();
+            }
+        };
+        reactContext.registerReceiver(wiredHeadsetReceiver, wiredHeadSetFilter);
+
+        // Detect Bluetooth device changes
+        bluetoothManager.start();
+
+    }
+
+    /**
+     * Helper method to set the output route to a Bluetooth device.
+     * @param enabled true if Bluetooth should use used, false otherwise.
+     */
+    private void setBluetoothAudioRoute(boolean enabled) {
+        if (enabled) {
+            audioManager.startBluetoothSco();
+            audioManager.setBluetoothScoOn(true);
+        } else {
+            audioManager.setBluetoothScoOn(false);
+            audioManager.stopBluetoothSco();
+        }
+    }
+
+    /**
+     * Helper method to trigger an audio route update when devices change.  It makes sure the
+     * operation is performed on the main thread.
+     */
+    private void onAudioDeviceChange() {
+        mainThreadHandler.post(mainThreadRunner);
+    }
+
+    /**
+     * Helper class to detect and handle Bluetooth device changes.  It monitors Bluetooth headsets
+     * being connected / disconnected and notifies the module about device changes when this occurs.
+     */
+    private class BluetoothManager {
```

Comment:
I don't like the `Manager` naming in general because the word is (1) generic and (2) overused and, consequently, doesn't mean much.